sdk/core/core-httpclients/src/xhrHttpClient.ts

```diff
@@ -0,0 +1,217 @@
+      typeof (x as ReadableStream).tee === "function",
+  );
+}
+
+/**
+ * A HttpClient implementation that uses XMLHttpRequest to send HTTP requests.
+ * @internal
+ */
+class XhrHttpClient implements HttpClient {
+  /**
+   * Makes a request over an underlying transport layer and returns the response.
+   * @param request - The request to be made.
+   */
+  public async sendRequest(request: PipelineRequest): Promise<PipelineResponse> {
+    const url = new URL(request.url);
+    const isInsecure = url.protocol !== "https:";
+
+    if (isInsecure && !request.allowInsecureConnection) {
+      throw new Error(`Cannot connect to ${request.url} while allowInsecureConnection is false.`);
+    }
+
+    const xhr = new XMLHttpRequest();
+
+    if (request.proxySettings) {
+      throw new Error("HTTP proxy is not supported in browser environment");
+    }
+
+    const abortSignal = request.abortSignal;
+    if (abortSignal) {
+      if (abortSignal.aborted) {
+        throw new AbortError("The operation was aborted.");
+      }
+
+      const listener = (): void => {
+        xhr.abort();
+      };
+      abortSignal.addEventListener("abort", listener);
+      xhr.addEventListener("readystatechange", () => {
+        if (xhr.readyState === XMLHttpRequest.DONE) {
+          abortSignal.removeEventListener("abort", listener);
+        }
+      });
+    }
+
+    addProgressListener(xhr.upload, request.onUploadProgress);
+    addProgressListener(xhr, request.onDownloadProgress);
+
+    xhr.open(request.method, request.url);
+    xhr.timeout = request.timeout;
+    xhr.withCredentials = request.withCredentials;
+    for (const [name, value] of request.headers) {
+      xhr.setRequestHeader(name, value);
+    }
+
+    xhr.responseType = request.streamResponseStatusCodes?.size ? "blob" : "text";
+
+    const body = typeof request.body === "function" ? request.body() : request.body;
+    if (isWebReadableStream(body) || isNodeReadableStream(body)) {
+      throw new Error("streams are not supported in XhrHttpClient.");
+    }
+
+    xhr.send(body === undefined ? null : body);
+
+    if (xhr.responseType === "blob") {
+      return new Promise((resolve, reject) => {
+        handleBlobResponse(xhr, request, resolve, reject);
+        rejectOnTerminalEvent(request, xhr, reject);
+      });
+    } else {
+      return new Promise(function (resolve, reject) {
+        xhr.addEventListener("load", () =>
+          resolve({
+            request,
+            status: xhr.status,
+            headers: parseHeaders(xhr),
+            bodyAsText: xhr.responseText,
+          }),
+        );
+        rejectOnTerminalEvent(request, xhr, reject);
+      });
+    }
+  }
+}
+
+function handleBlobResponse(
+  xhr: XMLHttpRequest,
+  request: PipelineRequest,
+  res: (value: PipelineResponse | PromiseLike<PipelineResponse>) => void,
+  rej: (reason?: any) => void,
+): void {
+  xhr.addEventListener("readystatechange", () => {
```

Comment:
Are there scenarios where we end up with multiple listeners and need to remove listeners / cleanup?

Reply:
Won't fix. Exact copy of xisting code.